lib/Analysis/DependenceAnalysis.cpp

```diff
@@ -31,17 +33,60 @@ static void checkMemrefDependence(SmallVectorImpl<Operation *> &memoryOps,
       if (source == destination)
         continue;
 
+      // Initialize the dependence list for this destination.
+      if (results.count(destination) == 0)
+        results[destination] = SmallVector<MemoryDependence>();
+
+      // Look for for inter-iteration dependences on the same memory location.
       MemRefAccess src(source);
       MemRefAccess dst(destination);
       FlatAffineValueConstraints dependenceConstraints;
       SmallVector<DependenceComponent, 2> depComps;
       DependenceResult result = checkMemrefAccessDependence(
           src, dst, depth, &dependenceConstraints, &depComps, true);
 
-      if (results.count(destination) == 0)
-        results[destination] = SmallVector<MemoryDependence>();
-
       results[destination].emplace_back(source, result.value, depComps);
+
+      // Also consider intra-iteration dependences on the same memory location.
+      if (src != dst)
+        continue;
+
+      // Look for the common loop body that src and dst share. If there is none,
+      // there is nothing more to do.
+      Block *commonBlock = source->getBlock();
+
+      while (!isa<AffineForOp>(commonBlock->getParentOp()) &&
+             commonBlock->getParentOp())
+        commonBlock = commonBlock->getParentOp()->getBlock();
+
+      if (!commonBlock)
+        continue;
```

Comment:
Ah no, scratch that, I think we really have to determine the nearest common block, otherwise this check won't work in the presence of nested `affine.if`s. Consider `A` and `B` accessing the same memref in:
```
for {
  if {
    if {
      A;
    }
    B;
  }
}
```
If we then query `commonBlock->findAncestorOpInBlock` for `A` and `B`, this would return the outer `if`-op both times, and neither `isBeforeInBlock` the other.

Reply:
Good point, I'll add a test for this scenario and adjust accordingly.